### ISSUE TITLE
Collision objects not publishing properly with PlanningSceneInterface as class member

### ISSUE THREAD

**Issue by swilcock0** [ros-planning/moveit2#407]
### Is there a method to "publish" planning_scene_interface changes from move_group? 

Currently I've been using this code below to create a subscriber which adds collision bricks to the scene, as part of a larger file. They appear in RVIZ, and until recently it was working well. Recently however the objects are being passed through. Republishing the objects through the motion planning plugin in RVIZ makes the planner recognise them properly.

UPDATE:
See next comment for link to a working demonstration code for this problem

```cpp
  
#include <streambuf>
#include <string>
#include <cmath>
#include <iostream>
#include <fstream>

#include <rclcpp/rclcpp.hpp>
#include <rclcpp/time_source.hpp>

#include <ament_index_cpp/get_package_share_directory.hpp>

#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit/planning_scene_interface/planning_scene_interface.h>

#include <moveit_msgs/msg/display_robot_state.hpp>
#include <moveit_msgs/msg/display_trajectory.hpp>
#include <moveit_msgs/msg/attached_collision_object.hpp>
#include <moveit_msgs/msg/collision_object.hpp>

#include <geometry_msgs/msg/pose.h>
#include <geometry_msgs/msg/pose_stamped.h>

namespace ph = std::placeholders;
using namespace std::chrono_literals;

auto SUCCESS = moveit::planning_interface::MoveItErrorCode::SUCCESS;

// Parameters
const std::string PLANNING_GROUP = "panda_arm";

const std::string node_name = "move_group_interface_tutorial";

static const rclcpp::Logger logger_ = rclcpp::get_logger(node_name);

class moveServer : public rclcpp::Node
{
public:
  moveServer()
    : Node(node_name, rclcpp::NodeOptions().automatically_declare_parameters_from_overrides(true))
    , move_group_(std::shared_ptr<rclcpp::Node>(std::move(this)), PLANNING_GROUP)
    , planning_scene_interface_()
  {
    brick_num = 0;
    attached = false;

    // Config moveit params
    this->move_group_.setPlanningTime(5.0);
    this->move_group_.setMaxVelocityScalingFactor(1.0);
    this->move_group_.setMaxAccelerationScalingFactor(1.0);
    this->move_group_.setNumPlanningAttempts(2);
    this->move_group_.allowReplanning(true);

    // Configure callback groups for multithreading
    callback_group_movement_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);

    auto sub_opt = rclcpp::SubscriptionOptions();
    sub_opt.callback_group = callback_group_movement_;

    subscriber_move_with_brick = this->create_subscription<geometry_msgs::msg::Pose>(
        "move_with_brick", 10, std::bind(&moveServer::callbackMoveWithBrick, this, ph::_1), sub_opt);
  }

  /// Move group interface for the robot
  moveit::planning_interface::MoveGroupInterface move_group_;

  /// Planning scene interface
  moveit::planning_interface::PlanningSceneInterface planning_scene_interface_;

private:
  bool pick()
  {
    // pose_target currently unused, hardcoded the brick in
    RCLCPP_INFO(logger_, "            PICKING");
    RCLCPP_INFO(logger_, "-----------------------------");
    geometry_msgs::msg::Pose brick_pose;

    brick_pose.position.x = 0;
    brick_pose.position.y = 0.4;
    brick_pose.position.z = 0 + 0.065 / 2;
    brick_pose.orientation.x = -1;
    brick_pose.orientation.y = 0.0;
    brick_pose.orientation.z = 0.0;
    brick_pose.orientation.w = 0.0;

    this->addBrick(brick_pose.position.x, brick_pose.position.y, brick_pose.position.z, brick_pose.orientation.x,
                   brick_pose.orientation.y, brick_pose.orientation.z, brick_pose.orientation.w);
    geometry_msgs::msg::Pose pre_grasp_pose;
    geometry_msgs::msg::Pose grasp_pose;
    geometry_msgs::msg::Pose post_grasp_pose;
    grasp_pose = brick_pose;
    grasp_pose.position.z += 0.065 / 2;

    pre_grasp_pose = grasp_pose;
    pre_grasp_pose.position.z += 0.25;
    post_grasp_pose = pre_grasp_pose;

    bool success;
    double fraction = 1.0;

    success = this->moveToPose(pre_grasp_pose);

    success = this->moveToPose(grasp_pose);
    if (success)
    {
      if (this->move_group_.attachObject("brick" + std::to_string(this->brick_num - 1)))
      {
        attached = true;
      }
      success = this->moveToPose(post_grasp_pose);
    }

    return success;
  }

  bool place(geometry_msgs::msg::Pose pose_target)
  {
    RCLCPP_INFO(logger_, "            PLACING");
    RCLCPP_INFO(logger_, "-----------------------------");
    geometry_msgs::msg::Pose pre_place_pose;
    geometry_msgs::msg::Pose place_pose = pose_target;
    geometry_msgs::msg::Pose post_place_pose;

    place_pose.position.z += 0.065;

    pre_place_pose = place_pose;
    pre_place_pose.position.z += 0.1;
    post_place_pose = pre_place_pose;

    bool success = false;
    double fraction = 1.0;

    success = this->moveToPose(pre_place_pose);

    success = this->moveToPose(place_pose);
    if (success)
    {
      if (this->move_group_.detachObject("brick" + std::to_string(this->brick_num - 1)))
      {
        attached = false;
      }
      fraction = this->moveToPose(post_place_pose);
    }

    return success;
  }

  void addBrick(double x, double y, double z, double x_o, double y_o,
                double z_o, double w)
  {
    moveit_msgs::msg::CollisionObject collision_object;
    collision_object.header.frame_id = this->move_group_.getPlanningFrame();

    // The id of the object is used to identify it.
    collision_object.id = "brick" + std::to_string(this->brick_num);
    this->brick_num = this->brick_num + 1;

    // Define a box to add to the world.
    shape_msgs::msg::SolidPrimitive primitive;
    primitive.type = primitive.BOX;
    primitive.dimensions.resize(3);
    primitive.dimensions[0] = 0.215;
    primitive.dimensions[1] = 0.103;
    primitive.dimensions[2] = 0.065;

    // Define a pose for the box (specified relative to frame_id)
    geometry_msgs::msg::Pose brick_pose;

    brick_pose.position.x = x;
    brick_pose.position.y = y;
    brick_pose.position.z = z;
    brick_pose.orientation.x = x_o;
    brick_pose.orientation.y = y_o;
    brick_pose.orientation.z = z_o;
    brick_pose.orientation.w = w;

    collision_object.primitives.push_back(primitive);
    collision_object.primitive_poses.push_back(brick_pose);
    collision_object.operation = collision_object.ADD;

    std::vector<moveit_msgs::msg::CollisionObject> collision_objects;
    collision_objects.push_back(collision_object);
    brick_id_vector.push_back(collision_object.id);

    // Now, let's add the collision object into the world
    RCLCPP_INFO(logger_, "Adding brick #%u into the world at %f %f %f {%f %f %f %f}", this->brick_num - 1, x, y, z, x_o, y_o, z_o, w);
    planning_scene_interface_.applyCollisionObjects(collision_objects);
  }

  void callbackMoveWithBrick(const geometry_msgs::msg::Pose::SharedPtr pose_ptr)
  {
    // Pick up a brick from a set location, then move it to the requested position
    auto pose_target = *pose_ptr;
    if (attached == false)
    {
      this->pick();
    }

    if (attached == true)
    {
      this->place(pose_target);
    }
  }

  bool moveToPose(geometry_msgs::msg::Pose pose_target)
  {
    // clearPathConstraints();
    this->move_group_.clearPoseTarget();
    this->move_group_.setStartState(*this->move_group_.getCurrentState(5));

    RCLCPP_INFO(logger_, "Targeting pose");
    this->move_group_.setPoseTarget(pose_target);

    moveit::planning_interface::MoveGroupInterface::Plan my_plan;

    bool success = (this->move_group_.plan(my_plan) == moveit::planning_interface::MoveItErrorCode::SUCCESS);
    if (success){
      this->move_group_.execute(my_plan);
    }

    RCLCPP_INFO(logger_, "Pose goal %s ", success ? "SUCCEEDED" : "FAILED");

    return success;
  }

  rclcpp::Subscription<geometry_msgs::msg::Pose>::SharedPtr subscriber_move_with_brick;

  std::vector<std::string> brick_id_vector;

  rclcpp::CallbackGroup::SharedPtr callback_group_movement_;

  bool attached;
  int brick_num;
};



int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);

  auto move_group_class = std::make_shared<moveServer>();

  rclcpp::executors::MultiThreadedExecutor exec;
  exec.add_node(move_group_class);

  RCLCPP_INFO(logger_, "Motion server has been started.");

  exec.spin();

  rclcpp::shutdown();
  return EXIT_SUCCESS;
}
```

![image](https://user-images.githubusercontent.com/48917295/113205718-80ca6780-9266-11eb-9da9-efaee0c68a65.png)
^ End effector moving through the object

### Your environment
* ROS Distro: Foxy
* OS Version: 20.04
* Source build: Latest main

### Steps to reproduce
Attempt to run a planning_scene_interface in a class

### Expected behaviour
Collision geometry recognised

### Actual behaviour
Geometry only appears visually

**Comment by swilcock0** [comment]
I've written a quick demo to properly demonstrate my problem with this, found here [swilcock0/Demo_MoveGroupInterface_Bug](https://github.com/swilcock0/moveit2/tree/swilcock0/Demo_MoveGroupInterface_Bug).  

Running
`ros2 launch run_move_group run_move_group.launch.py`

and then
`ros2 launch run_move_group run_move_group_oop_interface.launch.py`

gives a subscriber to the geometry_msgs/Pose topic /move_with_brick. Sending (any pose) to this topic to create a collision object and move to it, the manipulator then proceeds to move into the object, despite the brick being applied to the planning_scene_interface.

Any help much appreciated! @JafarAbdi seems a likely candidate for understanding what is likely a silly mistake on my part.

**Comment by felixvd** [comment]
Probably not the solution, but you should keep `planning_scene_interface` alive as a class member instead of recreating it. It's unnecessary overhead at the very least.

**Comment by swilcock0** [comment]
> Probably not the solution, but you should keep `planning_scene_interface` alive as a class member instead of recreating it. It's unnecessary overhead at the very least.

Thanks Felix, I have changed this in the commit referenced above - as you say, it unfortunately doesn't change the situation

**Comment by swilcock0** [comment]
Anyone any thoughts on this? I'm using a hacky fix of sorts by getting and applying the planning scene, but it has to be called a lot

**Comment by vatanaksoytezer** [comment]
This seems to be closely related with #429. We think this behavior is an upstream issue, but did not find the root yet.

**Comment by JafarAbdi** [comment]
@swilcock0 Do you mind testing https://github.com/ros-planning/moveit2/pull/456 and see if it fixes the bug

**Comment by swilcock0** [comment]
Hi @JafarAbdi just to confirm, yes it does

**Comment by henningkayser** [comment]
Fixed by #464
